src/algebraic_topology/fundamental_groupoid/retraction.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2022 Mark Lavrentyev. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mark Lavrentyev
+-/
+import topology.homotopy.equiv
+import category_theory.equivalence
+import algebraic_topology.fundamental_groupoid.product
+
+/-!
+# Homotopic maps induce naturally isomorphic functors
+
+## Main definitions
+
+  - `is_retraction r` A continuous map `r : X → A` (where `A` is a subtype of `X`) is a retraction
+    when the restriction of r to A is the identity. Equivalently, composing r with the inclusion
+    `i : A → X` is just the identity map on A.
+
+  - `fundamental_groupoid_split_mono_of_top_inclusion` The induced map on `fundamental_groupoid X`
+    of the inclusion `i : A → X` is a split mono in the category Groupoid.
+  - `fundamental_groupoid_split_epi_of_top_retraction` The induced map on `fundamental_groupoid X`
+    of the retraction `r : X → A` is a split epi in the category Groupoid.
+-/
+
+noncomputable theory
+
+universe u
+
+open fundamental_groupoid
+open category_theory
+open fundamental_groupoid_functor
+
+open_locale fundamental_groupoid
+open_locale unit_interval
+
+
+def top_hom_of_continuous_map
+  {X Y : Type*} [topological_space X] [topological_space Y] (f : C(X, Y)) : X ⟶ Y := f
+
+
+section unbundled
+
+/-- We define `A ⊆ X` to be a topological subspace by defining the property `A_filter` picking
+elements of `A` out of `X`. This inherits the topology on `X` via `subtype.topological_space`. -/
+variables {X : Type*} {A_filter : X → Prop} [topological_space X]
+
+def inclusion : C(subtype A_filter, X) := ⟨subtype.restrict A_filter id⟩
```

Comment:
You should probably use `{s : set X}` which is defeq to `X → Prop` but has the advantage that you don't need to write `subtype` anymore because a set is coerced automatically to a type.

I wonder if we should generalize from a subspace to a [inducing](https://leanprover-community.github.io/mathlib_docs/topology/maps.html#inducing) map here...

Reply:
I've adjusted the stuff to use `set` instead of `subtype` everywhere. Re:embedding I hadn't looked at that too much, but just giving a quick look here seems like a reasonable thing to do. I'll try setting it up with `embedding` instead of the inclusion I defined in the file and see how much has to change... hopefully not a ton :).